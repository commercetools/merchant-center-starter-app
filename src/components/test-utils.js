// This file allows consumers of the application-shell to test
// components rendered by the application shell.
// They should
// import { render } from ""@commercetools-frontend/application-shell/test-utils"
// and then use it together with react-testing-library.
import React from 'react';
import { Router } from 'react-router-dom';
import { render as rtlRender } from 'react-testing-library';
import { createMemoryHistory } from 'history';
import { IntlProvider } from 'react-intl';
import { MockedProvider as ApolloMockProvider } from 'react-apollo/test-utils';
import { ConfigureFlopFlip } from '@flopflip/react-broadcast';
import memoryAdapter from '@flopflip/memory-adapter';
import {
  FetchUser,
  FetchProject,
} from '@commercetools-frontend/application-shell';
import { GRAPHQL_TARGETS } from '@commercetools-frontend/constants';
import ProjectQuery from './project.graphql';
import UserQuery from './user.graphql';

// Reset memoryAdapter after each test, so that the next test accepts the
// defaultFlags param.
// This could also be moved into setup-test-framework, not sure which
// location is better for it.
afterEach(memoryAdapter.reset);

// TODO: define proper response data
const createDefaultMocks = ({ projectKey }) => [
  {
    request: {
      query: UserQuery,
      variables: { target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND },
    },
    result: {
      data: {
        user: {
          id: 'u1',
          email: '',
          gravatarHash: '',
          firstName: '',
          lastName: '',
          language: 'en',
          numberFormat: '',
          timeZone: '',
          launchdarklyTrackingId: '',
          launchdarklyTrackingGroup: '',
          launchdarklyTrackingTeam: '',
          defaultProjectKey: '',
          projects: {
            total: 1,
            results: [
              {
                name: 'p1',
                key: 'p1',
                suspension: {
                  isActive: false,
                },
                expired: false,
              },
            ],
          },
        },
      },
    },
  },
  {
    request: {
      query: ProjectQuery,
      variables: {
        target: GRAPHQL_TARGETS.MERCHANT_CENTER_BACKEND,
        projectKey,
      },
    },
    result: {
      data: {
        project: {
          key: '',
          version: '',
          name: '',
          countries: '',
          currencies: '',
          languages: '',
          expiry: {
            isActive: false,
            daysLeft: null,
          },
          suspension: {
            isActive: false,
          },
          permissions: {
            canManageOrganization: true,
            canManageProject: true,
            canViewProjectSettings: true,
            canViewProducts: true,
            canManageProducts: true,
            canViewOrders: true,
            canManageOrders: true,
            canViewCustomers: true,
            canManageCustomers: true,
            canViewTypes: true,
            canManageTypes: true,
            canViewPayments: true,
            canManagePayments: true,
          },
          owner: {
            id: 'o1',
            name: '',
            createdAt: '',
            teams: {
              count: 0,
            },
          },
          settings: {
            id: '',
            baseSettings: {
              id: '',
              userProjectSettings: '',
              pluginPreset: null,
              orderStatesVisibility: {
                isPaymentStateHidden: false,
                isShipmentStateHidden: false,
                isOrderStateHidden: false,
              },
            },
            productSettings: '',
            currentProductSettings: '',
          },
        },
      },
    },
  },
];

const holdOn = (amount = 0) =>
  new Promise(resolve => setTimeout(resolve, amount));

// This function renders any component within the application context.
// The context is not completely set up yet, some things are missing:
//   - Redux
//   - Tracking on context
//   - Project information
//   - possibly more that I'm not aware of right now
//
//  We can add these things as we go and when we need them.
//
// Inspired by
// https://github.com/kentcdodds/react-testing-library-course/blob/2a5b1560656790bb1d9c055fba3845780b2c2c97/src/__tests__/react-router-03.js
// eslint-disable-next-line import/prefer-default-export
export const render = async (ui, options = {}) => {
  const { projectKey = 'test-1' } = options;
  const {
    // react-intl
    locale = 'en',
    // Apollo
    mocks = createDefaultMocks({ projectKey }),
    addTypename = false,
    // react-router
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
    // flopflip
    adpater = memoryAdapter,
    flags = {},
    // forwarding to react-testing-library
    ...renderOptions
  } = options;
  const rendered = {
    ...rtlRender(
      <IntlProvider locale={locale}>
        <ApolloMockProvider mocks={mocks} addTypename={addTypename}>
          <ConfigureFlopFlip adapter={adpater} defaultFlags={flags}>
            <Router history={history}>
              <FetchUser>
                {({ isLoading: isLoadingUser }) => {
                  if (isLoadingUser) {
                    return null;
                  }
                  return (
                    <FetchProject projectKey={projectKey}>
                      {({ isLoading: isLoadingProject }) => {
                        if (isLoadingProject) {
                          return null;
                        }
                        return ui;
                      }}
                    </FetchProject>
                  );
                }}
              </FetchUser>
            </Router>
          </ConfigureFlopFlip>
        </ApolloMockProvider>
      </IntlProvider>,
      renderOptions
    ),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
    // return also `projectKey` in case it's the default value
    projectKey,
  };
  // wait for the user query to finish
  await holdOn();
  // wait for the project query to finish
  await holdOn();
  return rendered;
};
