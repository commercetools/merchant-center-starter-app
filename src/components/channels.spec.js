import React from 'react';
import { Provider as StoreProvider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
// import { fireEvent, waitForElement } from 'react-testing-library';
import { createMiddleware } from '@commercetools-frontend/sdk';
import { render } from './test-utils';
import ChannelsList from './channels-list';

const testProjectKey = 'test-1';

let mockedResponsesByUri = {};

const mockedResponse = ({ uri }) => {
  const response = mockedResponsesByUri[uri]();
  if (!response) throw new Error(`No mock found for uri ${uri}`);
  return response;
};

jest.mock('@commercetools-frontend/sdk/utils');
jest.mock('@commercetools/sdk-client', () => ({
  createClient: () => ({
    execute: mockedResponse,
  }),
}));

const mockStore = configureStore([
  thunk,
  createMiddleware({
    getCorrelationId: () => 1,
    getProjectKey: () => testProjectKey,
  }),
]);
const store = mockStore();

const mockQuery = (key, valueResolver) => {
  mockedResponsesByUri = {
    ...mockedResponsesByUri,
    [key]: valueResolver,
  };
};

const renderWithMockedStore = (ui, options) =>
  render(<StoreProvider store={store}>{ui}</StoreProvider>, options);

describe('rendering', () => {
  it('should render channels list', async () => {
    // Define the mock for the channels query (using the SDK)
    mockQuery(`/${testProjectKey}/channels?limit=25&offset=0`, () =>
      // Do not return anything for now
      Promise.resolve({ body: {} })
    );

    const { container } = await renderWithMockedStore(
      <ChannelsList projectKey={testProjectKey} />,
      { projectKey: testProjectKey }
    );

    expect(container.firstChild).toMatchSnapshot('Loading spinner');
  });
});
