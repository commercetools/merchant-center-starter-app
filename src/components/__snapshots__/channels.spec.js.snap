// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should render channels list: Loading spinner 1`] = `
<div
  class="inset-mod_m__RrUZk"
>
  <div
    class="stack-mod_m__3s4LT stack-mod_stretch__1Ck4r"
  >
    <h2
      class=""
    >
      <span>
        Channels
      </span>
    </h2>
    <div
      class="inline-mod_s__3wvxr inline-mod_center__1oQM5"
    >
      <div
        class="loading-spinner-mod_container-large__3fibW"
      >
        <svg
          class="loading-spinner-mod_svg-large__R2yVR"
          viewBox="0 0 40 40"
        >
          <path
            class="loading-spinner-mod_circle__3HSgV"
            d="M20.201,5.169c-8.254,0-14.946,6.692-14.946,14.946c0,8.255,6.692,14.946,14.946,14.946 s14.946-6.691,14.946-14.946C35.146,11.861,28.455,5.169,20.201,5.169z M20.201,31.749c-6.425,0-11.634-5.208-11.634-11.634c0-6.425,5.209-11.634,11.634-11.634c6.425,0,11.633,5.209,11.633,11.634C31.834,26.541,26.626,31.749,20.201,31.749z"
          />
          <path
            class="loading-spinner-mod_pointer__Ipsdk"
            d="M26.013,10.047l1.654-2.866c-2.198-1.272-4.743-2.012-7.466-2.012h0v3.312h0 C22.32,8.481,24.301,9.057,26.013,10.047z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
